### ISSUE TITLE
Final output is upcasted to float32 even though I used mixed_precision="fp16"

### ISSUE THREAD

**Issue by antoinedelplace** [huggingface/accelerate#3620]
### System Info

```Shell
- `Accelerate` version: 1.7.0
- Platform: Linux-6.1.123+-x86_64-with-glibc2.35
- `accelerate` bash location: /usr/local/bin/accelerate
- Python version: 3.11.13
- Numpy version: 2.0.2
- PyTorch version: 2.6.0+cu124
- PyTorch accelerator: CUDA
- System RAM: 12.67 GB
- GPU type: Tesla T4
- `Accelerate` default config:
	Not found
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [x] My own task or dataset (give details below)

### Reproduction

Final output is upcasted to float32 even though I used mixed_precision="fp16". I don't understand why.

Minimal example:
```
import torch
import torch.nn as nn
from accelerate import Accelerator
from torch import amp

class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Conv2d(3, 3, kernel_size=3, stride=1, padding=1)

    def forward(self, x):
        x = self.layer(x)
        print("x2", x.dtype, x.device)
        return x

accelerator = Accelerator(mixed_precision="fp16")
model = Net()
model = accelerator.prepare(model)
model.eval()

x = torch.rand((3, 1024, 1024), dtype=torch.float16, device="cuda")
print("x1", x.dtype, x.device)

with torch.no_grad(), amp.autocast(device_type='cuda', dtype=torch.float16):
    x = model(x)
    print("x3", x.dtype, x.device)

```

Output is :
```
x1 torch.float16 cuda:0
x2 torch.float16 cuda:0
x3 torch.float32 cuda:0
```

Can you help please? Am I doing something wrong?

### Expected behavior

Expected output is :
```
x1 torch.float16 cuda:0
x2 torch.float16 cuda:0
x3 torch.float16 cuda:0
```

**Comment by antoinedelplace** [comment]
It seems like this is called when the model forward is trigger :
https://github.com/huggingface/accelerate/blob/bee04f1b019061120d5fe08a440b0374b17d2202/src/accelerate/utils/operations.py#L818

How to disable the convert_outputs_to_fp32?

**Comment by antoinedelplace** [comment]
Maybe an option is needed here for the output dtype ?

https://github.com/huggingface/accelerate/blob/bee04f1b019061120d5fe08a440b0374b17d2202/src/accelerate/accelerator.py#L1567

**Comment by SunMarc** [comment]
The computation is done in fp16 but we convert the output to fp32 as this is more stable when calculating the loss 
https://github.com/huggingface/accelerate/pull/134

If you feel like you still prefer to keep the output in fp16 when training , we can add an arg for that.

**Comment by S1ro1** [comment]
I think this is desired behaviour, you don't really want to have output in fp16 even when training, atleast I haven't heard of that.
